### ISSUE TITLE
"Improve heuristic to determine whether to evaluate a static variable's initializer" broke compiling Qt Declarative

### ISSUE THREAD

**Issue by mstorsjo** [llvm/llvm-project#74070]
The commit 53a24c33f09b81b8f009afbabd05f7086db3f288, broke compiling of Qt Declarative (observed with version 5.15), where compilation now fails with this error:
```
In file included from jit/qv4baselineassembler.cpp:45:
jit/qv4assemblercommon_p.h:253:42: error: integer value -1 is outside the valid range of values [0, 7] for the enumeration type 'RegisterID' [-Wenum-constexpr-conversion]
  253 |     static const RegisterID NoRegister = RegisterID(-1);
      |                                          ^
```
The surprising thing is that this error only pops up when compiling with debug info enabled - without debug info, it still compiles fine.

The issue can be reproduced with this attached file, [enum-range-preproc.zip](https://github.com/llvm/llvm-project/files/13526389/enum-range-preproc.zip), compiled with `clang -target i686-w64-mingw32 -c -g enum-range-preproc.cpp`. Without the `-g` parameter, it successfully compiles.

CC @Michael137 @shafik @AaronBallman

**Comment by mstorsjo** [comment]
This issue can also be reproduced with this reduced source:
```c++
class a {
  void b();
  void *c;
};
typedef enum {} h;
class d {
  static const h e = h(-1);
};
class f : d {
public:
  void g();
};
void a::b() { reinterpret_cast<f *>(c)->g(); }
```

**Comment by Michael137** [comment]
If I understood https://reviews.llvm.org/D131528 correctly, the "out of range enum value is UB" error only triggers in constexpr contexts. So presumably without debug-info, that enum cast isn't in a constexpr context. But with [53a24c33f09b81b8f009afbabd05f7086db3f288](https://github.com/llvm/llvm-project/commit/53a24c33f09b81b8f009afbabd05f7086db3f288) we started evaluating the initialiser as a constant expression, hence the error now is allowed to trigger.

**Comment by shafik** [comment]
Yeah, generating debug info should not change the semantics of the program. Maybe you need to use `EvaluateAsInitializer` instead of `EvaluateAsConstantExpr`.

CC @cor3ntin who probably knows the ins and out here better than I do

**Comment by mstorsjo** [comment]
Thanks!

For reference, it seems like this issue has been avoided further upstream in QtDeclarative in newer versions, in https://github.com/qt/qtdeclarative/commit/636481a31110f1819efaf6500b25fbc395854311. But despite that, we shouldn't indeed change semantics based on whether debug info is enabled.

**Comment by Michael137** [comment]
> Yeah, generating debug info should not change the semantics of the program. Maybe you need to use `EvaluateAsInitializer` instead of `EvaluateAsConstantExpr`.
> 
> CC @cor3ntin who probably knows the ins and out here better than I do

Will do that then, thanks

There's a chance we'll revert the patch because of other debug-info related concerns, will keep this thread updated

**Comment by mstorsjo** [comment]
Ping, any update on either reverting this to unbreak building older Qt Declarative, or getting a fix for this issue?

**Comment by Michael137** [comment]
Decided revert the change that depends on this for now: https://github.com/llvm/llvm-project/pull/74580

**Comment by Michael137** [comment]
Reverted the changes, let me know if you're still seeing issues

**Comment by mstorsjo** [comment]
> Reverted the changes, let me know if you're still seeing issues

Thanks, things build successfully for me now again.
